backend/src/Tabulators/IRV.ts

```diff
@@ -0,0 +1,241 @@
+
+                    if (topRemainingRank == 0) {
+                        // set initial top rank
+                        topRemainingRankIndex = candidate.index
+                        topRemainingRank = ballot[candidate.index]
+                        isOverVote = false
+                    }
+                    else if (ballot[candidate.index] < ballot[topRemainingRankIndex]) {
+                        // set new top rank
+                        topRemainingRankIndex = candidate.index
+                        topRemainingRank = ballot[candidate.index]
+                        isOverVote = false
+                    }
+                    else if (ballot[candidate.index] === ballot[topRemainingRankIndex]) {
+                        // multiple top ranks, mark as over vote
+                        isOverVote = true
+                    }
+                }
+            })
+            if (isOverVote) {
+                // ballot is overvote
+                overVoteCount += 1
+            }
+            if (topRemainingRank === 0 || isOverVote) {
+                // ballot is exhausted
+                exhaustedVoteCount += 1
+                exhaustedVotes.push(...activeVotes.splice(i, 1))
+            }
+            else {
+                // give vote to top rank candidate
+                roundVoteCounts[topRemainingRankIndex] += 1
+            }
+        }
+        results.voteCounts.push(roundVoteCounts)
+        results.exhaustedVoteCounts.push(exhaustedVoteCount)
+        results.overVoteCounts.push(overVoteCount)
+        let voteThreshold = activeVotes.length / 2
+        
+        // get max number of votes
+        let remainingCandidatesIndexes = remainingCandidates.map(c => c.index)
+        let maxVotes = Math.max(...roundVoteCounts.filter((c, i) => remainingCandidatesIndexes.includes(i)))
+        if (maxVotes >= voteThreshold) {
+            // candidate meets the threshold
+            // get index of winning candidate
+            let winningCandidateIndex = roundVoteCounts.indexOf(maxVotes)
+            // add winner, remove from remaining candidates
+            results.elected.push(summaryData.candidates[winningCandidateIndex])
+            remainingCandidates = remainingCandidates.filter(c => !results.elected.includes(c))
+            // TODO: Some STV stuff
+        }
+        else {
+            // find candidate with least votes and remove from remaining candidates
+            let lastPlaceCandidateIndex = roundVoteCounts.indexOf(Math.min(...roundVoteCounts.filter((c, i) => remainingCandidatesIndexes.includes(i))))
+            remainingCandidates = remainingCandidates.filter(c => c !== summaryData.candidates[lastPlaceCandidateIndex])
+        }
+    }
+    // Sort data in order of candidate placements
+    // results.summaryData = sortData(summaryData, results.elected.concat(results.tied).concat(results.other))
+    return results
+}
+
+function getIRVBallotValidity(ballot: ballot) {
```

Comment:
The front end could use this information too when people are submitting their ballots. Would it be better to expose this api for consistency, or to reimplement this on the front end?

Reply:
Easiest would be to reimplement it. Plus you could add the "Potential over vote" warning to it.